pkg/controller/parameters.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+)
+
+func buildParameters(kubeClient kubernetes.Interface, namespace string, parametersFrom []v1alpha1.ParametersFromSource, parameters *runtime.RawExtension) (map[string]interface{}, error) {
+	params := make(map[string]interface{})
+	if parametersFrom != nil {
+		for _, p := range parametersFrom {
+			fps, err := fetchParametersFromSource(kubeClient, namespace, &p)
+			if err != nil {
+				return nil, err
+			}
+			for k, v := range fps {
+				if _, ok := params[k]; ok {
+					return nil, fmt.Errorf("conflict: duplicate entry for parameter %s", k)
+				}
+				params[k] = v
+			}
+		}
+	}
+	if parameters != nil {
+		pp, err := unmarshalRawParameters(parameters.Raw)
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range pp {
+			if _, ok := params[k]; ok {
+				return nil, fmt.Errorf("conflict: duplicate entry for parameter %s", k)
+			}
+			params[k] = v
+		}
+	}
+	return params, nil
+}
+
+func fetchParametersFromSource(kubeClient kubernetes.Interface, namespace string, parametersFrom *v1alpha1.ParametersFromSource) (map[string]interface{}, error) {
+	var params map[string]interface{}
+	if parametersFrom.SecretKeyRef != nil {
+		data, err := fetchSecretKeyValue(kubeClient, namespace, parametersFrom.SecretKeyRef)
+		if err != nil {
+			return nil, err
+		}
+		p, err := unmarshalJSON(data)
+		if err != nil {
+			return nil, err
+		}
+		params = p.(map[string]interface{})
+
+	}
+	return params, nil
+}
+
+func unmarshalRawParameters(in []byte) (map[string]interface{}, error) {
+	parameters := make(map[string]interface{})
+	if len(in) > 0 {
+		if err := yaml.Unmarshal(in, &parameters); err != nil {
+			return parameters, err
+		}
+	}
+	return parameters, nil
+}
+
+func unmarshalJSON(in []byte) (interface{}, error) {
```

Comment:
It might be a little confusing that you can only specify YAML inline right now instead of also in a secret key, but I'm prepared to live with that for now and sort it out later.